modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -3393,6 +3395,9 @@ else if (task instanceof ForceRebalanceExchangeTask) {
                     }
 
                     if (assignsMap != null && rebTopVer.equals(NONE)) {
+                        Map<Integer, IgniteInternalFuture<GridDhtPreloaderAssignments>> futAssigns =
+                            cctx.preloader().preloadAsync(cnt, exchFut, assignsMap);
```

Comment:
I think this should be done much earlier, and replace `generateAssignments` with the result of a new method which returns `Map<Integer, IgniteInternalFuture<GridDhtPreloaderAssignments>>`.
I think we can move assignments generation to PartitionPreloadManager.
I think we can leave the finalizing rebalance process as it is and make the PartitionPreloadManager as "partition preloading function" only.
The index rebuild procedure may be build-in in `GridDhtPreloader`.

wdyt?
should we discuss it closely?

Reply:
Done, thanks.